pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,526 @@ package service
 
 import (
 	"context"
+	"math"
 	"reflect"
+	"strconv"
+	"sync/atomic"
+	"time"
 
+	"github.com/vdaas/vald/internal/errgroup"
 	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/k8s/configmap"
+	"github.com/vdaas/vald/internal/k8s/decoder"
+	"github.com/vdaas/vald/internal/k8s/job"
+	mpod "github.com/vdaas/vald/internal/k8s/metrics/pod"
+	"github.com/vdaas/vald/internal/k8s/pod"
+	"github.com/vdaas/vald/internal/k8s/statefulset"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/safety"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/config"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/model"
 )
 
+const (
+	qualifiedNamePrefix string = "rebalancer.vald.vdaas.org/"
+)
+
+// Rebalancer represents the rebalancer interface.
 type Rebalancer interface {
-	Start(context.Context) (<-chan error, error)
+	Start(ctx context.Context) (<-chan error, error)
 }
 
-type rebalancer struct{}
+type rebalancer struct {
+	podName      string
+	podNamespace string
+
+	jobs           atomic.Value
+	jobName        string
+	jobNamespace   string
+	jobTemplate    atomic.Value
+	jobTemplateKey string // config map key of job template
+
+	agentName         string
+	agentNamespace    string
+	agentResourceType config.AgentResourceType
+	pods              atomic.Value
+	podMetrics        atomic.Value
+
+	jobConfigmapName string
+
+	statefulSets atomic.Value
+
+	leaderElectionID string
 
-func New(opts ...Option) (dsc Rebalancer, err error) {
+	rcd           time.Duration // reconcile check duration
+	eg            errgroup.Group
+	ctrl          k8s.Controller
+	tolerance     float64
+	rateThreshold float64
+	decoder       *decoder.Decoder
+}
+
+// NewRebalancer initialize job, configmap, pod, podMetrics, statefulset, replicaset, daemonset reconciler.
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
 	r := new(rebalancer)
-	for _, opt := range append(defaultOpts, opts...) {
+
+	for _, opt := range append(defaultRebalancerOpts, opts...) {
 		if err := opt(r); err != nil {
 			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 		}
 	}
+
+	job, err := job.New(
+		job.WithControllerName("job rebalancer"),
+		job.WithNamespaces(r.jobNamespace),
+		job.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		job.WithOnReconcileFunc(func(jobList map[string][]job.Job) {
+			log.Debugf("[reconcile Job] length Joblist: %d", len(jobList))
+			jobs, ok := jobList[r.jobName]
+			if ok {
+				r.jobs.Store(jobs)
+			} else {
+				r.jobs.Store(make([]job.Job, 0))
+				log.Infof("job not found: %s", r.jobName)
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := configmap.New(
```

Comment:
Is this reconcile loop really needed?
It seems that it only reads a `jobTemplate` from a configmap in this loop, and it is passed to `createJob` through `genJobTpl`.
I think it is not so different from mounting a configmap to the pod. Using a reconcile loop just makes the code complex.

Reply:
I'm not sure but is it easy to bundle the annotations or labels that are come from rebalance controller logic without using jobTemplates?

Reply:
In the current code, you're using `r.ctrl` for creating a job.
https://github.com/vdaas/vald/blob/4f80a872266f55652462a998181413cdae4be295/pkg/rebalancer/storage/controller/service/rebalancer.go#L383-L386
`cm` is just used to read a string data of `vald-rebalancer-storage-job-config`.

Reply:
@vankichi I think what rinx means is that mount the configmap to the pod without using reconciler .
e.g.  mount configmap as a env variable:
https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#define-a-container-environment-variable-with-data-from-a-single-configmap

Reply:
than you.
I fixed: https://github.com/vdaas/vald/pull/1187/commits/3007060163c7c736ca0882f8622b4874fae10b5a